### ISSUE TITLE
How do you extract the control points in different scenes? 

### ISSUE THREAD

**Issue by Jacobew** [ybarancan/STSU#15]
Hi, I wonder how did you extract the control points in different scenes including crossroad and straightway? I can't find the details in the paper.  Could you please explain that?
In a straight way, there is no obvious way to determine the distance between each sampled point, did you experiment on different settings?

**Comment by ybarancan** [comment]
Hi,
The centerlines are given in the dataset in global coordinates. So the ground truth is just given. We fit splines to the list of points per centerline as given in the dataset. You can check make_labels.py files and nusc_utils.get_centerlines for details. If you are asking how the method estimates the control points for a given image, it is a matter of monocular distance estimation training.

**Comment by Jacobew** [comment]
Thanks for the quick reply! 
I see, STSU just uses the original sampled points provided by the dataset. Then, what is the length per centerline segment? I think it matters a lot in the optimization difficulties. 
BTW, how do you predict the parameter t for a Bezier curve? Never mind if I'm missing some details. :)

**Comment by ybarancan** [comment]
In our definition a centerline segment is the curve between two consecutive 'divergence' points. So whenever there is a point with more than one leaving road or more than one entering road. Thus, the centerline segments have different lengths. The nice thing about using splines or Bezier curves is that it can represent curves with any length.  The only hyperparameter is the number of control points which I think is what you refer as t. We chose that as 3 by some heuristics like the fact that a centerline segment tends to be not too complicated. 2 control points can only represent a line so the next smallest number is 3.

**Comment by Jacobew** [comment]
Sorry for the misunderstanding, for the t I mean the scalar parameter that parameterizes the curve B(t) and t \in [0, 1]. 
Is t also a hyperparameter or is it a prediction?

**Comment by ybarancan** [comment]
I see. t essentially renders the curve. We use it for metric result calculations and polyline loss. And we set it to 100. So t becomes linspace(0,100). This gives you 100 points (not equidistant in euclidean sense) on the curve. Changing the number of points shouldn't change much either in loss or metrics but linear spacing is important.

**Comment by Jacobew** [comment]
I'm afraid there is still some misunderstandings.😂
For the t parameter, you've mentioned in Section 3.1 that t lies in [0,1] (not 100?).
<img width="877" alt="image" src="https://user-images.githubusercontent.com/26053586/222963171-7e8d898f-b2ae-4c6f-bb04-9a8fbd4342b7.png">

**Comment by ybarancan** [comment]
No it is the same thing. A given t value [0,1] gives you a point on the curve. We want a rendered curve. So we have 100 t values linearly spaced in [0,1]. That way from 3 control points you get 100 rendered curve points. This t parameter is not actually a part of the method since the method estimates and is trained by the control points. The image you sent is for 2 control points Bezier curve and always gives a line as indicated in the title Linear Bezier curves. We use 3 points. In the paper, in Lane graph representation section, you will see how we use parameter t.

**Comment by Jacobew** [comment]
Aha, I got it. Thanks for your explanation!
